### DISCUSSION TITLE
Performance improvement opportunities marshalling/unmarshalling

### DISCUSSION THREAD

**Discussion by jo-me** [woven-planet/go-zserio#138] [Ideas]
Hi,

when looking at the generated code there is some room for improvement when it comes to performance of some functions.
For example, there are quite some duplicate or unused variables, missing slice initializations and so on. Nothing seriously wrong but noticeable when looking at pprof cpu profiles.

The places I noticed can be easily solved by adapting the code generator a little bit or the core library functions (e,g, array decode) but some things require a deeper understanding of zserio (which i don't currently have). 

- Would you be open to accept a PR to tweak some places in the code? 

- Is performance optimization something that you are interested in for your use case?

- Do you maybe already have some things in mind where there are low-hanging fruit when it comes to performance / memory optimzations?

Thanks,

Jochen Mehlhorn <jochen.mehlhorn@mercedes-benz.com>, Mercedes-Benz Tech Innovation GmbH

[Provider Information](https://github.com/Daimler/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Comment by wichert** [comment]
Hi @jo-me! We would be very open to a PR with improvements. I've enabled discussions for this repository to give us a place to discuss things - that might be better than using the issue tracker for questions.

We are always interested in performance optimisation. Currently we do not have any Go zserio code in our performance-critical code, so we are not planning any work there ourselves at this moment.

**Reply by jo-me** [reply]
Thanks @wichert, I've created a PR https://github.com/woven-planet/go-zserio/pull/140 with some low-hanging fruit changes.

I think improving array marshalling/unmarshalling probably yields the most benefits as pprof says that it spends the most CPU there.

When looking at the generated code, some things probably can be optimized further, but adapting the code templates for that part is really difficult due to the different data types that the templates need to handle. I don't have an overview over the different features and edge cases.

One thing that immediately jumps out though is the additional RawArray that is filled before creating the final data structure that is available in the associated struct. Is that something that is really necessary?
